clippy_lints/src/trait_bounds.rs

```diff
@@ -260,6 +260,9 @@ impl TraitBounds {
                     SpanlessTy { ty: p.bounded_ty, cx },
                     p.bounds.iter().collect::<Vec<_>>()
                 );
+                if let TyKind::Path(qpath) = p.bounded_ty.kind;
+                if format!("{}:", rustc_hir_pretty::qpath_to_string(&qpath))
```

Comment:
I've never seen the function qpath_to_string before, and is never used in clippy, but it's extremely handy here. Should I change this to something else? Is there a reason it's never used?

Reply:
I don't know about this specifc case, but I'm very sure, that using paths like this is prone to false positives. I would guess that this wouldn't hold for paths with reexported items, aliasses or some weird relative path like `duck::super::quack`.

Is there a reason why [`clippy_utils::is_from_proc_macro`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.is_from_proc_macro.html#) wouldn't work in this case? It's suppose to handle cases, were the spans are not marked as originating from a macro :)

cc @blyxyas this might be interesting to you and you might also have some thoughts :).

Reply:
We could likely implement `WithSearchPat` for `WherePredicate` relatively easily

Reply:
[`qpath_to_string`](https://doc.rust-lang.org/beta/nightly-rustc/src/rustc_hir_pretty/lib.rs.html#218-220) is a really expensive (it uses [`print_path`](https://doc.rust-lang.org/beta/nightly-rustc/src/rustc_hir_pretty/lib.rs.html#1603)) function, so I think really the best way to go here, as Centri3 said, is by implementing `WithSearchPat` on `TyKind`.
(I will still benchmark these PR so we are sure that this is a performance problem)

Reply:
As a kind of side note, the average time for testing `type_repetition_in_bounds` in this PR is about 190 millis (run 1000 times). While in `master`'s branch, it's 168 millis.

Is this significant? *Maybe...* (It would be a nice addition to have `WithSearchPath` implemented for more like `TyKind` or `Attribute` (useful for `allow_attributes`, which is bugged for this exact reason. But it isn't a deal-breaker))

Reply:
> As a kind of side note, the average time for testing `type_repetition_in_bounds` in this PR is about 190 millis (run 1000 times). While in `master'`s branch, it's 168 millis.

I'm curious what the difference would be if `is_from_proc_macro` was used instead.

> is by implementing `WithSearchPat` on `TyKind`

For this PR in particular I think we should implement both on `TyKind` and  `WherePredicate`, since that's what's linted on anyway. afaict it would just need to search for `'` or the type's name (if it's `BoundPredicate` or `RegionPredicate`, respectively). `TyKind` (or `Ty` to be more specific) would likely

> or `Attribute`

This was actually already implemented in #10869 :)

Reply:
Hmm, I'm trying to implement it for `Ty` but am not sure what to use for `Path`... I'll commit the current implementation and hope someone can give their thoughts on it